### PR TITLE
WIP Remove charmstore.v5 dependency from Juju core

### PR DESCRIPTION
**Pull request by timClicks**
**In-progress**

----

## Description of change

*Please replace with a description about why this change is needed?*

## QA steps

*Please replace with how we can verify that the change works?*

## Documentation changes

*Please replace with any notes about how it affects current user workflow? CLI? API?* 

## Bug reference

*Please add a link to any bugs that this change is related to.*

### CODE REVIEW

File: apiserver/facades/client/application/application_test.go [hunk 1]
```diff
@@ -1317,21 +1280,29 @@ func (s *applicationSuite) TestSpecializeStoreOnDeployApplicationSetCharmAndAddC
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(results.Results, gc.HasLen, 1)
 	c.Assert(results.Results[0].Error, gc.IsNil)
-	c.Assert(repo.testMode, jc.IsTrue)
+	//c.Assert(repo.testMode, jc.IsTrue)
```
Comment (wallyworld) [inline apiserver/facades/client/application/application_test.go]: remove me
but this does test a key aspect of what this test was written for, which was to ensure that the test mode was set on the repo if so configured in the model
the patched repo should record the test mode and expose it for tests to check

File: apiserver/facades/client/application/application_test.go [hunk 2]
```diff
@@ -1317,21 +1280,29 @@ func (s *applicationSuite) TestSpecializeStoreOnDeployApplicationSetCharmAndAddC
...
+	modelConfig, err = s.Model.ModelConfig()
```
Comment (wallyworld) [inline apiserver/facades/client/application/application_test.go]: this looks like it duplicates what's above

File: apiserver/facades/client/application/application_test.go [hunk 3]
```diff
@@ -1317,21 +1280,29 @@ func (s *applicationSuite) TestSpecializeStoreOnDeployApplicationSetCharmAndAddC
...
+	c.Assert(modelConfig.TestMode(), jc.IsTrue)
 
 	// Check that the store's test mode is enabled when calling SetCharm.
 	curl, _ = s.UploadCharm(c, "trusty/wordpress-2", "wordpress")
 	err = s.applicationAPI.SetCharm(params.ApplicationSetCharm{
 		ApplicationName: "application",
 		CharmURL:        curl.String(),
 	})
-	c.Assert(repo.testMode, jc.IsTrue)
+	// c.Assert(repo.testMode, jc.IsTrue)
```
Comment (wallyworld) [inline apiserver/facades/client/application/application_test.go]: and here and below

File: apiserver/facades/controller/charmrevisionupdater/testing/suite.go [hunk 4]
```diff
@@ -5,17 +5,16 @@ package testing
...
 
-	jc "github.com/juju/testing/checkers"
 	gc "gopkg.in/check.v1"
 	"gopkg.in/juju/charm.v6"
 	"gopkg.in/juju/charmrepo.v3"
-	"gopkg.in/juju/charmrepo.v3/csclient"
-	"gopkg.in/juju/charmstore.v5"
 
+	jc "github.com/juju/testing/checkers"
+	// "github.com/juju/juju/apiserver/testing"
```
Comment (wallyworld) [inline apiserver/facades/controller/charmrevisionupdater/testing/suite.go]: remove me

File: apiserver/facades/controller/charmrevisionupdater/updater.go [hunk 5]
```diff
@@ -124,9 +124,9 @@ func retrieveLatestCharmInfo(st *state.State) ([]latestCharmInfo, error) {
 		return nil, errors.Trace(err)
 	}
 
-	var charms []charmstore.CharmID
-	var resultsIndexedApps []*state.Application
-	for _, application := range applications {
+	resultsIndexedApps := make([]*state.Application, len(applications))
```
Comment (wallyworld) [inline apiserver/facades/controller/charmrevisionupdater/updater.go]: no need for any of the changes in this file

File: apiserver/testing/fakecharmstore.go [hunk 6]
```diff
@@ -4,82 +4,485 @@
...
 	"gopkg.in/juju/charmrepo.v3"
 	"gopkg.in/juju/charmrepo.v3/csclient"
-	"gopkg.in/juju/charmstore.v5"
-	"gopkg.in/macaroon-bakery.v2-unstable/bakery/checkers"
-	"gopkg.in/macaroon-bakery.v2-unstable/bakerytest"
-	"gopkg.in/mgo.v2"
+	"gopkg.in/juju/charmrepo.v3/csclient/params"
+	"gopkg.in/macaroon.v1"
 
+	// jc "github.com/juju/testing/checkers"
```
Comment (wallyworld) [inline apiserver/testing/fakecharmstore.go]: remove me

File: apiserver/testing/fakecharmstore.go [hunk 7]
```diff
@@ -4,82 +4,485 @@
...
+	originalChannel := c.repo.channel
+	defer func() { c.repo.channel = originalChannel }()
+	return c.repo.WithChannel(channel).ResourceMeta(id, name, revision)
+}
+
+// ServerURL returns the empty string, as it is not accessing an actual charm store.
+//
+// ServerURL is part of the charmstore.CharmstoreWrapper interface
+func (c Client) ServerURL() string {
+	return ""
```
Comment (wallyworld) [inline apiserver/testing/fakecharmstore.go]: better to return a url string that parses, like "https://fake" or something

File: cmd/juju/application/bundle_test.go [hunk 8]
```diff
@@ -43,6 +39,9 @@ import (
 	"github.com/juju/juju/testcharms"
 	coretesting "github.com/juju/juju/testing"
 	"github.com/juju/juju/testing/factory"
+	"github.com/juju/loggo"
```
Comment (wallyworld) [inline cmd/juju/application/bundle_test.go]: move me

File: cmd/juju/application/bundle_test.go [hunk 9]
```diff
@@ -202,7 +201,7 @@ func (s *BundleDeployCharmStoreSuite) TestDeployBundleWithTermsSuccess(c *gc.C)
 		"terms1/0": "0",
 		"terms2/0": "1",
 	})
-	c.Assert(s.termsString, gc.Not(gc.Equals), "")
+	// c.Assert(s.termsString, gc.Not(gc.Equals), "")
```
Comment (wallyworld) [inline cmd/juju/application/bundle_test.go]: fix me

File: cmd/juju/application/deploy_test.go [hunk 10]
```diff
@@ -6,60 +6,51 @@ package application
...
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
 	"github.com/juju/juju/juju/testing"
 	"github.com/juju/juju/juju/version"
 	"github.com/juju/juju/jujuclient"
 	"github.com/juju/juju/jujuclient/jujuclienttesting"
 	"github.com/juju/juju/state"
 	"github.com/juju/juju/testcharms"
 	coretesting "github.com/juju/juju/testing"
+	"github.com/juju/juju/testing/factory"
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: move this block

File: cmd/juju/application/deploy_test.go [hunk 11]
```diff
@@ -529,6 +523,30 @@ func (s *CAASDeploySuiteBase) assertUnitsCreated(c *gc.C, expectedUnits map[stri
 	c.Assert(created, jc.DeepEquals, expectedUnits)
 }
 
+func (s *CAASDeploySuiteBase) fakeAPI() *fakeDeployAPI {
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: this doesn't need to be on the suite

File: cmd/juju/application/deploy_test.go [hunk 12]
```diff
@@ -990,117 +1028,33 @@ const (
...
-		IdentityLocation: s.discharger.Location(),
-		PublicKeyLocator: keyring,
-		TermsLocation:    s.termsDischarger.Location(),
-	}
-	handler, err := charmstore.NewServer(db, nil, "", params, charmstore.V5)
-	c.Assert(err, jc.ErrorIsNil)
-	s.handler = handler
-	s.srv = httptest.NewServer(handler)
-	c.Logf("started charmstore on %v", s.srv.URL)
+func (s *charmStoreSuite) SetUpSuite(c *gc.C) {
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: not needed

File: cmd/juju/application/deploy_test.go [hunk 13]
```diff
@@ -990,117 +1028,33 @@ const (
...
+	s.JujuConnSuite.SetUpSuite(c)
+}
 
-	s.client = csclient.New(csclient.Params{
-		URL:      s.srv.URL,
-		User:     params.AuthUsername,
-		Password: params.AuthPassword,
-	})
+func (s *charmStoreSuite) TearDownSuite(c *gc.C) {
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: ditto

File: cmd/juju/application/deploy_test.go [hunk 14]
```diff
@@ -990,117 +1028,33 @@ const (
...
-		}})
-		return actualNewCharmStoreClient(client, s.srv.URL)
-	})
-
-	// Point the Juju API server to the charm store testing server.
-	s.PatchValue(&csclient.ServerURL, s.srv.URL)
+	s.client = apiservertesting.NewClient(apiservertesting.NewRepository())
 }
 
 func (s *charmStoreSuite) TearDownTest(c *gc.C) {
```
Comment (wallyworld) [inline cmd/juju/application/deploy_test.go]: and this is no longer needed

File: cmd/juju/application/deploy.go [hunk 15]
```diff
@@ -260,7 +265,10 @@ func NewDeployCommandForTest(newAPIRoot func() (DeployAPI, error), steps []Deplo
 }
 
 // NewDeployCommand returns a command to deploy applications.
-func NewDeployCommand() modelcmd.ModelCommand {
+func NewDeployCommand(
```
Comment (wallyworld) [inline cmd/juju/application/deploy.go]: We don't want to use any arg here and have to update all the call sites to NewDeployCommand(nil. nil)
The default values are nil and the ForTest() variant passes in non-nil values

**Comment by timClicks** [comment]
Closing in favour of #9977

**Pull request closed without merging**